src/util/animation/AnimationBase.ts

```diff
@@ -0,0 +1,167 @@
+   */
+  private startTime!: number;
+
+  /**
+   * Constructor
+   * Since both `byValue` and `endValue` are accepted in subclass options
+   * and are populated with defaults if missing, we defer to `byValue` and
+   * ignore `endValue` to avoid conflict
+   */
+  constructor({
+    startValue,
+    byValue,
+    duration = 500,
+    delay = 0,
+    easing = defaultEasing,
+    onStart = noop,
+    onChange = noop,
+    onComplete = noop,
+    abort = defaultAbort,
+    target,
+  }: Partial<TAnimationBaseOptions<T> & TAnimationCallbacks<T>> &
+    Required<Omit<TAnimationValues<T>, 'endValue'>>) {
+    this.duration = duration;
+    this.delay = delay;
+    this.easing = easing;
+    this._onStart = onStart;
+    this._onChange = onChange;
+    this._onComplete = onComplete;
+    this._abort = abort;
+    this.target = target;
+
+    this.startValue = startValue;
+    this.byValue = byValue;
+    this.value = this.startValue;
+    this.endValue = this.calculate(this.duration).value;
+  }
+
+  get state() {
+    return this._state;
+  }
+
+  /**
+   * Calculate the current value based on the easing parameters
+   * @param timeElapsed in ms
+   * @protected
+   */
+  protected abstract calculate(timeElapsed: number): {
+    value: T;
+    changeRatio: number;
+  };
+
+  start() {
+    const firstTick: FrameRequestCallback = (timestamp) => {
+      if (this._state !== 'pending') return;
+      this.startTime = timestamp || +new Date();
+      this._state = 'running';
+      this._onStart();
+      this.tick(this.startTime);
+    };
+
+    this.register();
+
+    // setTimeout(cb, 0) will run cb on the next frame, causing a delay
+    // we don't want that
+    if (this.delay > 0) {
+      setTimeout(() => requestAnimFrame(firstTick), this.delay);
+    } else {
+      requestAnimFrame(firstTick);
+    }
+  }
+
+  private tick: FrameRequestCallback = (t) => {
+    const durationMs = (t || +new Date()) - this.startTime;
+    const boundDurationMs = Math.min(durationMs, this.duration);
+    this.durationRatio = boundDurationMs / this.duration;
+    const { value, changeRatio } = this.calculate(boundDurationMs);
+    this.value = Array.isArray(value) ? (value.slice() as T) : value;
+    this.valueRatio = changeRatio;
+
+    if (this._state === 'aborted') {
+      return;
+    } else if (this._abort(value, this.valueRatio, this.durationRatio)) {
+      this._state = 'aborted';
+      this.unregister();
+    } else if (durationMs >= this.duration) {
+      const endValue = this.endValue;
+      this.durationRatio = this.valueRatio = 1;
+      this._onChange(
+        (Array.isArray(endValue) ? endValue.slice() : endValue) as T,
+        this.valueRatio,
+        this.durationRatio
+      );
+      this._state = 'completed';
+      this._onComplete(endValue, this.valueRatio, this.durationRatio);
+      this.unregister();
+    } else {
+      this._onChange(value, this.valueRatio, this.durationRatio);
+      requestAnimFrame(this.tick);
+    }
+  };
```

Comment:
There is clear indication on MDN that fat arrows shouldn't be used for class methods.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions

Unless there is a strong reason around not following that advice, this should be a standard method and an explicit bind in the constructor, would work correctly from the prototype AND bound out of context.

Reply:
I read that post and wasn't convinced that was what it said...
But I don't mind binding

Reply:
done